model-server/src/main/kotlin/org/modelix/model/server/store/IStoreClient.kt

```diff
@@ -20,29 +20,164 @@ import kotlinx.coroutines.coroutineScope
+        val map = getAll(keys.toSet())
+        return keys.map { map[it] }
+    }
+
+    override fun getAll(keys: Set<String>): Map<String, String?> {
+        val fromRepository = client.getAll(keys.map { it.withRepoScope() }.toSet()).mapKeys { it.key.key }
+        if (getRepositoryId() == null) return fromRepository
+
+        // Existing databases may have objects stored without information about the repository.
+        // Try to load these legacy entries.
+        val missingKeys = fromRepository.entries.asSequence().filter { it.value == null }.map { ObjectInRepository.global(it.key) }.toSet()
+        val fromGlobal = client.getAll(missingKeys).mapKeys { it.key.key }
+
+        return fromRepository + fromGlobal
+    }
+
+    override fun getAll(): Map<String, String?> {
+        throw UnsupportedOperationException()
+        // return client.getAll().filterKeys { it.repositoryId == null || it.repositoryId == repositoryId?.id }.mapKeys { it.key.key }
+    }
+
+    override fun put(key: String, value: String?, silent: Boolean) {
+        client.put(key.withRepoScope(), value, silent)
+    }
+
+    override fun putAll(entries: Map<String, String?>, silent: Boolean) {
+        client.putAll(entries.mapKeys { it.key.withRepoScope() }, silent)
+    }
+
+    override fun listen(key: String, listener: IGenericKeyListener<String>) {
+        client.listen(key.withRepoScope(), RepositoryScopedKeyListener(listener))
+    }
+
+    override fun removeListener(key: String, listener: IGenericKeyListener<String>) {
+        client.removeListener(key.withRepoScope(), RepositoryScopedKeyListener(listener))
+    }
+
+    override fun generateId(key: String): Long {
+        return client.generateId(key.withRepoScope())
+    }
+
+    override fun <T> runTransaction(body: () -> T): T {
+        return client.runTransaction(body)
+    }
+
+    override fun close() {
+        client.close()
+    }
+}
+
+fun IGenericStoreClient<ObjectInRepository>.forRepository(repository: RepositoryId): RepositoryScopedStoreClient {
+    return RepositoryScopedStoreClient(repository, this)
+}
+fun IGenericStoreClient<ObjectInRepository>.forGlobalRepository() = GlobalStoreClient(this)
+fun IGenericStoreClient<ObjectInRepository>.forContextRepository() = ContextScopedStoreClient(this)
+fun <R> IStoreClient.withGlobalRepository(body: () -> R): R = withRepository(null, body)
+fun <R> IStoreClient.withRepository(repository: RepositoryId?, body: () -> R): R {
+    assert(this is ContextScopedStoreClient || this is StoreClientAdapter && this.getRepositoryId() == repository) {
+        "Store is not context scoped: $this"
+    }
+    return ContextScopedStoreClient.CONTEXT_REPOSITORY.computeWith(repository, body)
+}
+suspend fun <R> IStoreClient.withGlobalRepositoryS(body: suspend () -> R): R = withRepositoryS(null, body)
+
+suspend fun <R> IStoreClient.withRepositoryS(repository: RepositoryId?, body: suspend () -> R): R {
```

Comment:
What is that `S` suffix? Do we have that convention used somewhere else already?

Reply:
Renamed it to `withRepositoryInCoroutine`.